### ISSUE TITLE
Show created and update times separately on annotation cards in the sidebar

### ISSUE THREAD

**Issue by seanh** [hypothesis/client#1240]
# Problem

Currently only a single date is shown on annotation cards in the client (in the top right of the card). It doesn't say in the UI whether this is the annotation's creation or update time. In fact it's the update time, which leads to this undesirable behavior:

1. Create an annotation
2. Come back some time later (say a couple of months or years) and, noticing a small typo, edit your annotation and correct it
3. The date on your annotation changes to today, rather than the date when you originally wrote the annotation, making it look as if you just wrote it today

While this behavior is confusing, unexpected and undesirable in general, it's particularly acute when an instructor is using the client to grade a student's annotations in the LMS app. The creation date might be important if it relates to the submission deadline of the assignment, but if the student has edited the annotation then only the update date is shown.

# Solution

1. The main date shown on annotation cards should be their original creation date, not their update date
2. Secondarily, it should also be possible to see whether an annotation has been edited and if so when (at least the date of the most recent edit)

**Comment by dwhly** [comment]
Yay.  Never knew this.  Added design label for the 2nd part.

**Comment by dwhly** [comment]
Some further thoughts:
- Obviously switching the primary timestamp to reflect first creation is the most important change here.  That was a misfire in the original implementation.  
- Some sites do not show visual indication of updated time on cards.  Facebook is probably the key example here (although the version history is available from the menu on each card).
- Some sites do show update times, Reddit is a good example of this.  That info is shown just after the creation time stamp.  [An example](https://www.reddit.com/r/UberEATS/comments/9ohzxp/nyc_who_do_i_call_if_i_cant_get_my_bike_unlocked/e839t6q/) is "9 months ago - _edited just now_".
- Other sites (trying to find an example) show an updated reference lower down after the content body itself-- reflecting a practice that users sometimes self-enforce of including "Updated" text at the end of the content, or just above additional content added to posts that they've made previously).

Given the importance to teachers of knowing whether a student may have updated a submission after a deadline has passed, it seems important that we implement some form of indication, and that we indicate them separately.

Potential approaches
- The tooltip over the timestamp could include "updated" or "edited" info. (perhaps on a 2nd line?)  Currently it provides the full time info, which is helpful since the timestamp display is a truncated and simplified version of time.
- We provide a secondary timestamp after the primary one, like reddit, using the same simplified time format that we do in the first timestamp.  It could either be on the same line, or perhaps on a secondary line.  It appears most of the time we would have sufficient space to show it.  The second timestamp is not hyperlinked, but does have its own tooltip, so that the precise update time can be discovered.
- We provide the update time somewhere else on the card-- lower down-- either before the primary content or after it.

One question:
- If someone updates the content *immediately*, as in 30 seconds later-- should we still show the updated field?  Or only if the update time passes some threshold (1, 5, 10, 15 minutes? or if there is at least one reply to the annotation?).  It's likely that a large percentage of edits happen almost immediately, simply as a result of someone realizing they wanted to make a slight modification-- but not because of the passage of time, new information, or in response to a reply.  We may not need to draw the users attention to the edit if that's the case.  Possibly the distinct edit time could always be in the tooltip, but not called out in the actual UX unless t > 5 minutes, etc.

**Comment by lyzadanger** [comment]
I'm working on an in-browser prototype using our existing client UI components and am ready to collaborate on getting this done!

Disclaimers:

* This is still a sketchy in-browser prototype. Details like exact date formats and whatnot are still up in the air. e.g. Whether we use a relative date format such as “18m ago”.
* There are various states an annotation can have, which will need to be represented in this interface. The examples here—for now—only cover a subset.
* This prototype takes into account much of the existing comments here with a considerable influence from GitHub, e.g.:
![image](https://user-images.githubusercontent.com/439947/61977469-8add9280-afbc-11e9-8c02-5aa04df9b255.png)

* The term _edited_ means that the textual body of the annotation has been altered and re-saved (re-published) at any time since the creation of the annotation. 

The most immediate thing to remedy is that the timestamp displayed on the card’s header should be the created date. And we also need to surface some information about whether the annotation has been edited, and when the last edit was.

So:

An annotation that has never been edited:

![image](https://user-images.githubusercontent.com/439947/61977525-a779ca80-afbc-11e9-90b3-a315456eb4ce.png)

An annotation that has been edited at least once:

![image](https://user-images.githubusercontent.com/439947/61977532-af396f00-afbc-11e9-8dc9-1becc464aaee.png)

Menu disclosure exposes more details about annotation creation and edit dates:

![image](https://user-images.githubusercontent.com/439947/61977610-ea3ba280-afbc-11e9-8694-49cb3d9d2e7b.png)

We get the nice hover from the Menu component:

![image](https://user-images.githubusercontent.com/439947/61977689-1ce59b00-afbd-11e9-90bd-4b169144fb82.png)


Now I will attempt to answer some questions I think might arise:

* Why has a “**View annotation**” action been made into a menu item and put in a menu?

(This might be the most controversial thing here).

Right now, the timestamp performs double-service as a link to the annotation, but _you have to know that_. Discoverability seems iffy and I’d wager that it’s not immediately obvious to many (most?) users that the timestamp is linked. It’s not styled as a link and is visually quiet.

This prototype always renders a disclosure arrow next to the timestamp, communicating that it is an interactive element. The “view annotation” menu item clearly describes what it does and also has the added benefit of  using the consistent “external” icon, indicating that you’re about to be taken away from the current context.

Yes, it’s an extra click, but since the entire text of the timestamp area is part of the menu label, it’s easy to get at (click anywhere on the timestamp or the disclosure area to open the menu). 

Using our existing Menu component allows quick reuse of something without having to invent a new visual metaphor.

* Why does “View annotation” come first in the menu?

So that it’s always in the same place and immediately proximal for mouse click/tap. The length of the rest of the menu is dependent on whether the annotation has been edited or not—I didn’t want the “view annotation” action jumping around.

* What about the positioning of the “edited” text hint?

At present, the word “edited” comes _after_ the created timestamp on the card header. My first revision had this differently, see:

![image](https://user-images.githubusercontent.com/439947/61977745-4f8f9380-afbd-11e9-8a01-e52b985b47ac.png)

We’re dealing with a right-aligned element, and I’d prefer that the create-timestamps always align flush right and don’t “move around.” However, putting the word “edited” _before_ the timestamp seemed like it might lead to confusion; it reads as if the annotation was “edited [on] date”, which is not the case.

My next-step goal in this discussion is to build consensus around the idea of using a Menu-type component in this UI and the formatting of the timestamp we want to use when the Menu is not open; I feel like the contents of the menu and their details can be iterated on.

**Comment by xolotl** [comment]
In keeping with Hypothesis' principle to be "[International, by design](https://web.hypothes.is/principles/)" I'm hoping dates/times can either be fully localizable, but if not, perhaps displayed using a "little-endian", but easily-understandable-to-all non-USA-centric format like 4 Jul 1776 rather a "middle-endian" format like Jul 4 1776.

**Comment by lyzadanger** [comment]
> In keeping with Hypothesis' principle to be "International, by design" I'm hoping dates/times can either be fully localizable, but if not, perhaps displayed using a "little-endian", but easily-understandable-to-all non-USA-centric format like 4 Jul 1776 rather a "middle-endian" format like Jul 4 1776.

Yes, I'm glossing over a lot of that here, but it is already supported by a bunch of the underlying mechanisms. Pretend we're in the US for _now_, because we are, but yep!
